### PR TITLE
Whats the usage difference between EspMqttClient::new and EspMqttClient::new_cb ?

### PR DESCRIPTION
**Pull request by AZarbade**
**TLDR: I want to know when should I be using `EspMqttClient::new` and `EspMqttClient::new_cb` ?** Also I believe mqtt example should be updated.

[documentation reference](https://docs.esp-rs.org/esp-idf-svc/esp_idf_svc/mqtt/client/struct.EspMqttClient.html).
- using `EspMqttClient::new` returns `Result<(Self, EspMqttConnection), EspError>`
- while using `EspMqttClient::new_cb` returns `Result<Self, EspError>`

**Note**: I am using xtensa-esp32-.

Following the given [mqtt example](https://github.com/esp-rs/std-training/blob/main/intro/mqtt/exercise/examples/solution_publ.rs) I am assuming that I need to use `EspMqttClient::new_cb` version. Which has been working as expected. 
Here is my code snippet:
```
let mut mqtt_client = EspMqttClient::new_cb(&broker_url, &mqtt_config, move |_msg| {})?;
```

However, when I try to use the other varient `EspMqttClient::new` nothing is getting published to the mqtt broker, even though I haven't changed the publishing method.
Here is my code snippet:
```
let (mut mqtt_client, _mqtt_connection) = EspMqttClient::new(&broker_url, &mqtt_config)?;
```

**What am I supposed to do with `_mqtt_connection`?**

For reference, here is complete code snippet;
```
let mqtt_config = MqttClientConfiguration::default();
let mut mqtt_client = EspMqttClient::new_cb(&broker_url, &mqtt_config, move |_msg| {})?;

let payload: &[u8] = &[];
mqtt_client.publish("home/default", QoS::AtLeastOnce, true, payload)?;
```

**Suggestion:**
The example defines the new client as follows,
`
let mut client = EspMqttClient::new(&broker_url, &mqtt_config, move |_message_event| {})?;
`
but the function has been updated to use `EspMqttClient::new_cb` instead for `EspMqttClient::new`. 
Updated code:
`
let mut mqtt_client = EspMqttClient::new_cb(&broker_url, &mqtt_config, move |_message_event| {})?;
` 

PS: This is literally my first bug report/pull request, so please let me know if I need to provide any additional information or if there are any issues with my report.

### CODE REVIEW

**Comment by Vollbrecht** [comment]
First you can have a look at our official example [here](https://github.com/esp-rs/esp-idf-svc/blob/master/examples/mqtt_client.rs#L44). 
The technical difference between the two api's is that with the callback api, the things that need to be done will directly be called from within the thread that is executing the mqtt stack itself. This has pro's and con's. It maybe simpler to use as you don't have to fiddle with a messaging system, because you get everything directly handeld. One disadvantage is that you are executing in the mqtt stack itself, so you need to limit yourself with the work you are doing in the callback, so to not starve the system. The mqtt stack is executed in a higher priority thread.
With the "new" api your execution runs in either your main thread or in a separate thread.So running in general in a lower priority compared to the still ruining mqtt-stack thread. This rises to some awkard problems at the start, but in general decouples the system a bit.
